tools/generate.py

```diff
@@ -365,14 +395,68 @@ def _generate_full_name(prefix: str, name: str) -> str:
     return full_name
 
 
+def _normalize_string(s: str) -> str:
+    return re.sub(" +", " ", s)
+
+
+def _build_function_info_doc(
+    fullname: str,
+    args: list[str],
+    return_args: list[str],
+    return_names: list[str],
+    ignore_return_doc: bool = False,
+) -> str | None:
+    if not fullname in DOCS:
+        return None
+
+    args = [a.replace("*", "") for a in args]
+
+    doc = DOCS[fullname]
+    arg_doc: dict[str, str | None] = {}
+    for arg in args + return_names:
+        arg_doc[arg] = _normalize_string(DOCS.get(f"{fullname}.{arg}", ""))
+
+    return_doc = _normalize_string(DOCS.get(f"{fullname}.$return-value", ""))
+
+    # There are two conditions:
+    # 1: A return value and some / none out arguments,
+    #    in this case the length will be different and we should ignore the first return_args
+    # 2: No return value, in this case the length will be equal
+    add_one = len(return_args) != len(return_names)
+    if not ignore_return_doc:
+        return_doc = (
+            f"{return_args[0]}: {return_doc or 'Not documented'}\n    "
+            if add_one and not (len(return_args) == 1 and return_args[0] == "None")
+            else None
+        )
+    else:
+        return_doc = None
+
+    ret_args: dict[str, str] = {}
+    for i, arg in enumerate(return_names):
+        index = i + 1 if add_one else i
+        ret_args[arg] = return_args[index]
+
+    return f"""{doc}
+
+Parameters:
+    {'\n    '.join([f'{s}: {arg_doc[s]}' for s in args])}
+
+Returns:
+    {return_doc or ""}{'\n    '.join([f'{ret_args[s]}: {arg_doc[s]}' for s in return_names])}
```

Comment:
This appears to require Python ≥ 3.12, whereas `pyproject.toml` claims to support ≥ 3.7:

> SyntaxError: f-string expression part cannot include a backslash